crates/wasi-nn/src/backend/pytorch.rs

```diff
@@ -88,32 +101,36 @@ impl BackendExecutionContext for PytorchExecutionContext {
             .iter()
             .map(|&dim| dim as i64)
             .collect::<Vec<_>>();
-        self.1 = TchTensor::from_data_size(&input_tensor.data, &dimensions, kind);
+        self.inputs.push(TchTensor::from_data_size(
+            &input_tensor.data,
+            &dimensions,
+            kind,
+        ));
```

Comment:
This is almost there: we're still ignoring the index, though. What might need to happen is that we set up inputs as a `Vec<Option<TchTensor>>` filled with `None` and then set the right item based on the index. We are able to retrieve the number of inputs, right?

Reply:
Did you mean use the index as the position/index for the vector of tensors? That sounds good. I'll see if there's a way to determine the max number of inputs to the given model.

Reply:
Alright, so here's what I've changed -

- `set_input()` now uses the `id` passed to it. If it is `u32`, it assigns the tensor to the vector `inputs[id]`, assigning `None` if there are any empty spots below the given `id`. This will give the user flexibility to set input tensors non sequentially.
- `compute()` will check the vector for any `None` values, and give an error if present, before calling `forward_ts()`.
- There is no reliable way at this time to get max inputs for a model. However, assigning more inputs than available returns an error message detailing the max inputs the model expects at run time, so this is helpful for the user. The expectation, similar to other backends, is for the user to be aware of input size/shape etc.
- If `id` is a string, I've currently permitted only single input to be set, effectively ignoring the index. If more inputs are assigned, or if `u32` and `String` indexes are used together, we give out an error.

Let me know if this looks fine. Thanks.

Reply:
Yeah, looking at [727a4aa](https://github.com/bytecodealliance/wasmtime/pull/9234/commits/727a4aa31e54519371377a2796afe73719e745b2), I think that makes more sense. I was kind of hoping that [`CModule::named_parameters`](https://docs.rs/tch/latest/tch/struct.CModule.html#method.named_parameters) would return the list of inputs by name, but, if that's not the case, then let's just return an error for `Id::Name` side of things. In any case, we probably don't need `id_type`: if we get an `Id::Index` we know were to put it in the vector and if we get an `Id::Name` then we either (a) look up its index in `named_parameters` or (b) if that's not possible, return an error.

Reply:
I had added the Id:Name case because the wit test cases (`nn_wit_image_classification_pytorch` in this case) need a Name instead of Index to pass - [nn.rs](https://github.com/bytecodealliance/wasmtime/blob/80ed395e185b7092bb0b520b67150b7b984e1148/crates/test-programs/src/nn.rs#L43) and [wasi-nn.wit](https://github.com/rahulchaphalkar/wasmtime/blob/786b073664950ebcf3b4fe801826bc17ad3207e3/crates/wasi-nn/wit/wasi-nn.wit#L121). Although it looks like `set-input` might be going away anyway.